Product property building refactoring

0nko opened the pull request:
This PR moves all the product detail property-building logic out of the `ProductDetailFragment` into a separate `ProductDetailCardBuilder` class. All the property types are abstracted away from the Android views and the hierarchy is now generated inside the ViewModel. The view hierarchy is constructed using nested `RecyclerViews`, which allows for a lot of flexibility. The same approach was used in Stats refresh in WPAndroid and it worked fairly well.

This whole thing has a couple benefits:

- The UI is modelled using simple data classes, which means the Product detail view hierarchy can be unit tested. The previous implementation had a lot of decision logic based on the product state from the ViewModel, which was breaking the rule that views should only be responsible for displaying data.
- The `ProductDetailFragment` was becoming too large, reaching almost 1000 LOC. Now, every property type has its own model and ViewHolder class, which are easier to navigate.
- New property types can be added easily in a modular way.
- All property models can now be reused on screens other than Product details, such as Product variations. This was my main motivation for the refactoring. All we need is to have a RecyclerView and a CardBuilder.

I know this PR is pretty massive but it wasn't really possible to split it up. I'd like to merge it to `develop` because the chance for merge conflict is very high. With that in mind, I'd like to ask everyone to please smoke test the Product details screen thoroughly, so that no new bugs are introduced. I've tried be careful to follow the old code, but you know how it is... :)

Any suggestions welcome!

nbradbury commented:
Looks like the description is no longer rendering HTML. This description has bolded words but they're not showing up:

![pr](https://user-images.githubusercontent.com/3903757/81178153-c5298680-8f75-11ea-9f4d-83d8440b8a78.png)

In `develop` the bolding is working:

![Screenshot_1588768975](https://user-images.githubusercontent.com/3903757/81178175-cd81c180-8f75-11ea-85df-39ac7a57aabd.png)

0nko commented:
> Looks like the description is no longer rendering HTML

@nbradbury Fixed.

nbradbury commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/WCProductPropertyView.kt:
```diff
@@ -62,9 +62,14 @@ class WCProductPropertyView @JvmOverloads constructor(
     }
 
+    fun removeClickListener() {
+        this.isClickable = false
```
Should this also nullify the assigned click listener, or does that not matter?

nbradbury commented:
@0nko There's some very noticeable redrawing and refreshing when returning to product detail from product settings. The refreshing also happens after updating the product.

![redraw](https://user-images.githubusercontent.com/3903757/81212452-e94f8c80-8fa2-11ea-971c-83c5ed120c32.gif)

nbradbury commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/WCProductPropertyCardView.kt:
```diff
@@ -30,5 +35,24 @@ class WCProductPropertyCardView @JvmOverloads constructor(
             divider.show()
         }
+
+        propertiesRecyclerView.layoutManager = LinearLayoutManager(context, RecyclerView.VERTICAL, false)
```
We can probably skip this if the layout manager is already assigned.

nbradbury commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/WCProductPropertyCardView.kt:
```diff
@@ -30,5 +35,24 @@ class WCProductPropertyCardView @JvmOverloads constructor(
+    }
+
+    private fun loadData(data: List<ProductProperty>) {
+        val recyclerView: RecyclerView = view.findViewById(R.id.propertiesRecyclerView)
```
I think we can skip `findViewById` here and just refer to `propertiesRecyclerView`.

nbradbury commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/adapters/ProductPropertiesDiffCallback.kt:
```diff
@@ -0,0 +1,45 @@
+    override fun areItemsTheSame(oldItemPosition: Int, newItemPosition: Int): Boolean {
+        val newItem = newList[newItemPosition]
+        val oldItem = oldList[oldItemPosition]
+        return if (oldItem.type == newItem.type) {
```
I'm wondering if this `when` clause is necessary. Can't we just check `oldItem == newItem`?

0nko replied:
It's not. Good call.

nbradbury commented:
@0nko This all looks very slick! I left a few nitpicks in the comments, but I do think the refresh/redraw issue will need to be addressed before merging.

I'm also not sure why tests are failing, since as we said on Slack they're working for us.

0nko commented:
@nbradbury Thank you for the review! I've addressed your code comments and the update flashing should be fixed as well now. Ready for another pass.

Regarding the failing tests, I'm not really sure what to do. Maybe @oguzkocer could give us some hints? (TLDR; locally all unit tests are passing, but some are failing on CircleCI)

oguzkocer commented:
@0nko Can you try running `./gradlew test` and verify that all tests pass for you? I see the exact same failures that CI reports even though the tests pass when I run it from Android Studio.

0nko commented:
Thanks @oguzkocer. After hours of investigations it turned out it was caused by a feature flag being enabled in Debug mode only, which made some tests fail in the Release mode 🤦.

@nbradbury All the comments should be addressed now and the unit tests are passing. Ready for another round!
